File: src/api/notebook.ts
import { RequestFN } from "~/core/request-function";
import { decodeNotebook } from "~/decoders/notebook";
import { encodePeriod } from "~/encoders/period";
import { encodePronoteDate } from "~/encoders/pronote-date";
import { type SessionHandle, type Notebook, type Period, TabLocation } from "~/models";

export const notebook = async (session: SessionHandle, period: Period): Promise<Notebook> => {
  const request = new RequestFN(session, "PagePresence", {
    _Signature_: { onglet: TabLocation.Notebook },

    donnees: {
      periode: encodePeriod(period),

      DateDebut: {
        _T: 7,
        V: encodePronoteDate(period.startDate)
      },

      DateFin: {
        _T: 7,
        V: encodePronoteDate(period.endDate)
      }
    }
  });

  const response = await request.send();
  return decodeNotebook(response.data.donnees);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/notebook.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeNotebook
import ink.literate.pawnote.encoders.encodePeriod
import ink.literate.pawnote.encoders.encodePronoteDate
import ink.literate.pawnote.models.Notebook
import ink.literate.pawnote.models.Period
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun notebook(session: SessionHandle, period: Period): Notebook {
  val request =
      RequestFN(
          session.information,
          "PagePresence",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Notebook.code) }

                putJsonObject("donnees") {
                  put("periode", encodePeriod(period))

                  putJsonObject("DateDebut") {
                    put("_T", 7)
                    put("V", encodePronoteDate(period.startDate))
                  }

                  putJsonObject("DateFin") {
                    put("_T", 7)
                    put("V", encodePronoteDate(period.endDate))
                  }
                }
              }))

  val response = request.send()
  return decodeNotebook(
      Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject, session)
}
